### PR TITLE
Adopt Yarn for CI

### PR DESCRIPTION
**Pull request by nathanhammond**
- [x] Travis works.
- [x] Use environment `yarn` and `npm`.
- [X] [`npm` fallback works on 0.12](https://travis-ci.org/ember-cli/ember-cli/jobs/183669109)
- [X] [AppVeyor works.](https://ci.appveyor.com/project/embercli/ember-cli/build/11040)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -116,7 +117,8 @@
     "tree-sync": "^1.2.1",
     "uuid": "^3.0.0",
     "walk-sync": "^0.3.0",
-    "yam": "0.0.22"
+    "yam": "0.0.22",
+    "yarn": "latest"
```
Comment (stefanpenner) [inline package.json]: Curious what the rationale is for `latest`

File: tests/acceptance/addon-smoke-test-slow.js [hunk 2]
```diff
@@ -34,23 +34,18 @@ describe('Acceptance: addon-smoke-test', function() {
...
   });
 
   afterEach(function() {
     // Cleans up a folder set up on the other side of a symlink.
     fs.remove(path.join(addonRoot, 'node_modules', 'developing-addon'));
 
-    return cleanupRun(addonName).then(function() {
-      expect(dir(addonRoot)).to.not.exist;
-    });
+    cleanupRun(addonName);
```
Comment (stefanpenner) [inline tests/acceptance/addon-smoke-test-slow.js]: In distance past this was sync, in the not-so-distant past this was made async due to windows being unable to consistently synchronously remove large trees of files.

As such I am a tad nervous to change this, but if demonstrated that windows doesn't suffer from this issue. My nerves will be settled.

File: appveyor.yml [hunk 3]
```diff
@@ -15,25 +15,19 @@ branches:
...
   - '%APPDATA%\Roaming\bower'
 
-
 # Post-install test scripts.
 test_script:
   # Output useful info for debugging.
-  - npm version
+  - git rev-parse HEAD
+  - yarn version
   - cmd: npm run test-all
```
Comment (nathanhammond) [inline appveyor.yml]: Can't use `yarn` to run for Node.js 0.12. This is a hybrid configuration right now, intentionally.

File: package.json (hunk 1 above)
Comment (nathanhammond) [inline package.json]: I'm going to remove this.

File: .travis.yml [hunk 4]
```diff
@@ -16,26 +16,29 @@ branches:
...
+    - $HOME/.npm
+    - $HOME/.cache # includes bower's cache
 
 before_install:
+  # Once we no longer support Node.js 0.12 we can drop this entire section.
   # prevent the npm loading indicator
   - npm config --global set spin false
-  # if npm version is less 3.0.0 (lets say 1.x or 2.x) we should attempt to upgrade to 3
+  # if npm version is less than 3.0.0, upgrade to 3
   - if [[ $(npm -v | cut -d '.' -f 1) -lt 3 ]]; then npm i -g npm@^3; fi
```
Comment (nathanhammond) [inline .travis.yml]: I've set this up as a hybrid config for `npm` and `yarn`.

File: appveyor.yml (hunk 3 above)
Comment (nathanhammond) [inline appveyor.yml]: Oh, this is AppVeyor. Yes, this is a bug.

File: tests/acceptance/addon-smoke-test-slow.js (hunk 2 above)
Comment (nathanhammond) [inline tests/acceptance/addon-smoke-test-slow.js]: We'll find out in time. 😄

File: appveyor.yml [hunk 5]
```diff
@@ -28,7 +28,6 @@ cache:
 test_script:
   # Output useful info for debugging.
   - git rev-parse HEAD
-  - yarn --version
```
Comment (stefanpenner) [inline appveyor.yml]: is the version printed out already, or...

File: package.json (hunk 1 above)
Comment (nathanhammond) [inline package.json]: This has now been removed.

**Comment by nathanhammond** [comment]
Aiming to land this today having just done a rebase. I still probably have a few little things which aren't 100% that I'll need to sort through but we'll know shortly.

File: appveyor.yml (hunk 5 above)
Comment (stefanpenner) [inline appveyor.yml]: another good default by our friends working on `yarn`

**Comment by stefanpenner** [comment]
@nathanhammond +1 on merging when CI is green. As implemented everything looks very reasonable. If you need to make dramatic changes, and want a review just ask.

**Comment by rwjblue** [comment]
@homu r+

**Pull request merged**